### ISSUE TITLE
Access to TypedDict NotRequired fields is not checked inside with statement

### ISSUE THREAD

**Issue by unicodestrangely** [microsoft/pyright#7714]
**Describe the bug**
Access to TypedDict NotRequired fields is not checked inside with statement.

**Code or Screenshots**

```python
from typing import TypedDict, NotRequired

class MyDict(TypedDict):
    foo: str
    bar: NotRequired[int]

def get_bar(d: MyDict):
    print(d["bar"])  # = reportTypedDictNotRequiredAccess
    with open("/etc/hostname"):  # Can be any expression
        print(d["bar"])  # = no error
```

**VS Code extension or command-line**
Command-line, 1.1.358

**Comment by erictraut** [comment]
This is by design, so I don't consider this a bug.

This check is suppressed if used within a `try` block or a context manager block (which act as implicit `try` blocks in some cases). This was done in response to [complaints from pyright users](https://github.com/microsoft/pyright/issues/2378) who (rightly) pointed out that pyright was complaining about a legitimate use — and a common practice in Python.

**Comment by unicodestrangely** [comment]
Reconsider the design decision, please.

I don't think suppressing exceptions in `__exit__` is a common practice in Python. Even original issuer is willing to accept it's a poor coding practice.

Warning may be implicitly suppressed inside `with suppress(KeyError)`, but inside other `with` statements Pyright should warn about a potential runtime error (because it is a potential runtime error). If someone handles `KeyError` inside `__exit__` they should explicitly use `# pyright: ignore [...]` for this very uncommon case.

**Comment by erictraut** [comment]
I agree that this decision is worth reconsidering. I'm going to reopen and explore the backward compatibility impact of making this change. No promises at this point.

**Comment by unicodestrangely** [comment]
I don't understand what is so special with `NotRequired` so it needs to be special cased. It leads to potential errors without any gains. 
```python
from typing import NotRequired, TypedDict

class MyDict(TypedDict):
    bar: NotRequired[int]

def my_func(bar: int):
    print(bar + 1)

d: MyDict = {}
my_func(d["foo"])  # = reportGeneralTypeIssues
my_func(d["bar"])  # = reportTypedDictNotRequiredAccess
with open("/etc/hostname"):
    my_func(d["foo"])  # = reportGeneralTypeIssues
    my_func(d["bar"])  # = no error?
```

Argument that context manager may be used to suppress exceptions is not a particulary strong one.
```python
from typing import NotRequired, TypedDict

class MyDict(TypedDict):
    bar: NotRequired[int]

class LookMaNoExceptions:
    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        # Don't do this in real code
        return True

d: MyDict = {}
with LookMaNoExceptions():
    # Shouldn't all these errors be implicitly suppressed? (No, they shouldn't)
    1 + ""  # pyright error
    open(None)  # pyright error
    d["bar"]  # no error?
```

Errors should never pass silently. Unless explicitly silenced (with `# pyright: ignore`)

**Comment by erictraut** [comment]
I've decided to completely eliminate the exemption for both context managers and `try` blocks, which means this check will apply in both of these cases. The previous exemption was inconsistent with other behaviors in pyright, which assumes that exception handling is used only to handle exceptional conditions, not for anticipated control flow or type-related guards.

This change may create some pain for pyright users who were relying on this exemption.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.360, which I just published.
